include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -178,10 +184,17 @@ class pool_memory_resource final
    */
   block_type block_from_upstream(size_t size, cudaStream_t stream)
   {
-    void* p = upstream_mr_->allocate(size, stream);
-    block_type b{reinterpret_cast<char*>(p), size, true};
-    upstream_blocks_.emplace_back(b);  // TODO: with C++17 use version that returns a reference
-    return b;
+    RMM_LOG_INFO("[A][Stream {}][Upstream {}B]", reinterpret_cast<void*>(stream), size);
+
+    try {
+      void* p = upstream_mr_->allocate(size, stream);
+      block_type b{reinterpret_cast<char*>(p), size, true};
+      upstream_blocks_.emplace_back(b);  // TODO: with C++17 use version that returns a reference
+      return b;
+    } catch (std::exception const& e) {
+      RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE]", reinterpret_cast<void*>(stream), size);
+      throw e;
```

Comment:
This would throw a copy of the base exception instead of throwing the same exception object, i.e., you'd lose any derived object. Instead, rethrow the same exception object. 
```suggestion
      throw;
```